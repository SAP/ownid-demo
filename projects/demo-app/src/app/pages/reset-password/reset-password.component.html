<div class="container" *ngIf="pwrt$ | async as pwrt">
  <a class="logo" routerLink="/"></a>

  <div class="content">
    <div class="title">Reset Password</div>

    <div class="description">Enter your new password.</div>

    <ownid type="recover" [data]="{pwrt:pwrt}" (onRecover)="onRecover($event)"></ownid>

    <div class="or">OR</div>

    <form class="own-form" [formGroup]="form" (ngSubmit)="onSubmit(pwrt)">
      <div class="errors" *ngIf="errors">{{errors}}</div>

      <input class="own-input" type="password" placeholder="New Password" formControlName="password">
      <input class="own-input" type="password" placeholder="Confirm Password" formControlName="confirmPassword">

      <button class="own-button" type="submit">Set new Password</button>
    </form>
  </div>
</div>


