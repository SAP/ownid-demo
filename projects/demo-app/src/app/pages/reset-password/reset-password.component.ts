import { Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { Observable } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { GigyaService } from '../../services/gigya.service';

@Component({
  selector: 'login',
  templateUrl: './reset-password.component.html',
  styleUrls: ['./reset-password.component.scss']
})
export class ResetPasswordComponent {
  form: FormGroup;

  errors: string | null = null;

  pwrt$: Observable<string>;

  constructor(
    actRoute: ActivatedRoute,
    formBuilder: FormBuilder,
    private gigyaService: GigyaService,
    private router: Router,
  ) {
    this.pwrt$ = actRoute.queryParamMap.pipe(
      filter((params) => !!params.get('pwrt')),
      map((params) => params.get('pwrt') as string),
    );

    this.form = formBuilder.group({
      password: ['', [Validators.required]],
      confirmPassword: ['', [Validators.required]],
    });
  }

  onSubmit(passwordResetToken: string) {
    if (this.form.valid) {
      this.errors = null;

      const params = {
        passwordResetToken,
        newPassword: this.form.value.password,
      }

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.gigyaService.resetPassword(params, (data: any) => {
        if (data.status === 'FAIL') {
          this.errors = data.errorDetails;
          return;
        }

        this.router.navigateByUrl('/login');
      });
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onRecover($event: any) {
    console.log($event);
    this.router.navigateByUrl('/login');
  }
}
