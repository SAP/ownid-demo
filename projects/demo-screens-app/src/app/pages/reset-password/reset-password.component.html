<demo-header></demo-header>

<demo-footer></demo-footer>


