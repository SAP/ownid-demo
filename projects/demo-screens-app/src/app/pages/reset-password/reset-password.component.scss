:host {
  background: #E5E5E5;
}

.logo {
  margin: 0 auto;
}

.title {
  font-size: 20px;
  line-height: 32px;
  font-weight: bold;
}

.description {
  font-size: 14px;
  line-height: 20px;
  margin-bottom: 24px;
}

.container {
  max-width: 480px;
  width: calc(100% - 16px);
  margin: 48px auto 0;
}

.content {
  margin-top: 40px;
  padding: 24px 24px 32px;

  background: #FFFFFF;
  box-shadow: 0 8px 16px rgba(4, 172, 167, 0.16), 0 0 2px rgba(4, 172, 167, 0.16);
  border-radius: 6px;
}
