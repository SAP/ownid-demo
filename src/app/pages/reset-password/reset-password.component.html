<section class="sign-body">
  <div class="container">
    <div class="call-2-action">
      <h1 class="headline">
        For the best experience use Own Id for login
      </h1>
      <h2 class="sub-title">
        Scan QR code with your phone and proceed with out any logins and
        passwords
      </h2>
    </div>
    <div>
      <mat-card class="sign-section">
        <mat-card-title>
          Set New password or use OwnId to recover your account
        </mat-card-title>

        <ng-container *ngIf="pwrt$ | async as pwrt">
          <ownid type="recover" [data]="{pwrt:pwrt}" (onRecover)="onRecover($event)"></ownid>

          <div class="or">OR</div>

          <form [formGroup]="form" (ngSubmit)="onSubmit(pwrt)">
            <p>
              <span class="errors" *ngIf="errors">{{errors}}</span>
            </p>
            <p>
              <mat-form-field class="form-field">
                <mat-label>Set New Password</mat-label>
                <input
                    matInput
                    placeholder="New Password"
                    formControlName="password"
                    type="password"
                    required
                />
              </mat-form-field>
            </p>
            <p>
              <mat-form-field class="form-field">
                <mat-label>Confirm Password</mat-label>
                <input
                    matInput
                    placeholder="Confirm Password"
                    formControlName="confirmPassword"
                    type="password"
                    required
                />
              </mat-form-field>
            </p>
            <button mat-flat-button color="primary" class="form-field" type="submit">Set new Password</button>
          </form>
        </ng-container>
      </mat-card>
    </div>
  </div>
</section>





