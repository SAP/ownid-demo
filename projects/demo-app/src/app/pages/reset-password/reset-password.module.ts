import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { OwnidModule } from '../../shared/ownid/ownid.module';
import { ResetPasswordRoutingModule } from './reset-password-routing.module';
import { ResetPasswordComponent } from './reset-password.component';

@NgModule({
  declarations: [ResetPasswordComponent],
  imports: [
    CommonModule,
    ResetPasswordRoutingModule,
    ReactiveFormsModule,
    OwnidModule,
  ]
})
export class ResetPasswordModule { }
