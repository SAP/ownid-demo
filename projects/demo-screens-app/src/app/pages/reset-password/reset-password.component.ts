import { Component, ElementRef } from '@angular/core';
import { FormBuilder, FormGroup } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { BehaviorSubject, Observable } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { ResetPasswordCommand } from './commands/reset-password.command';
import { AppStore } from '../../app.store';

@Component({
  selector: 'login',
  templateUrl: './reset-password.component.html',
  styleUrls: ['./reset-password.component.scss'],
})
export class ResetPasswordComponent {
  form: FormGroup;

  errors$: BehaviorSubject<string | null>;

  pwrt$: Observable<string>;

  private observer: MutationObserver;

  constructor(
    actRoute: ActivatedRoute,
    formBuilder: FormBuilder,
    appStore: AppStore,
    private resetPasswordCommand: ResetPasswordCommand,
    private router: Router,
    private elementRef: ElementRef,
  ) {
    this.pwrt$ = actRoute.queryParamMap.pipe(
      filter((params) => !!params.get('pwrt')),
      map((params) => params.get('pwrt') as string),
    );

    this.form = formBuilder.group({
      password: ['', []],
      confirmPassword: ['', []],
    });

    this.errors$ = appStore.formError$;

    // @ts-ignore
    this.observer = new MutationObserver(() => window.ownidWidget?.recalculatePosition());

    this.observer.observe(this.elementRef.nativeElement, {
      attributes: false,
      childList: true,
      subtree: true,
    });
  }

  ngOnDestroy() {
    this.observer.disconnect();
  }

  onSubmit(passwordResetToken: string) {
    if (this.form.valid) {
      this.resetPasswordCommand.execute({ passwordResetToken, password: this.form.value.password });
    }
  }

  onRecover() {
    this.router.navigateByUrl('/login');
  }

  onError(errorMessage: string) {
    this.errors$.next(errorMessage);
  }
}
