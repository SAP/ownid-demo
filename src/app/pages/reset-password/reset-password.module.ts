import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { ReactiveFormsModule } from "@angular/forms";
import { MatCardModule } from "@angular/material/card";
import { MatInputModule } from "@angular/material/input";
import { MatButtonModule } from "@angular/material/button";

import { OwnidModule } from "../../shared/ownid/ownid.module";
import { ResetPasswordRoutingModule } from "./reset-password-routing.module";
import { ResetPasswordComponent } from "./reset-password.component";

@NgModule({
  declarations: [ResetPasswordComponent],
  imports: [
    CommonModule,
    ResetPasswordRoutingModule,
    ReactiveFormsModule,
    OwnidModule,
    MatCardModule,
    MatInputModule,
    MatButtonModule
  ]
})
export class ResetPasswordModule {}
