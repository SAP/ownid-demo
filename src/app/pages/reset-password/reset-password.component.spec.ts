import { async, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { ActivatedRoute, Params, Router } from '@angular/router';
import { FormBuilder } from '@angular/forms';
import { of } from 'rxjs';

import { GigyaService } from '../../services/gigya.service';
import { ResetPasswordComponent } from './reset-password.component';

describe('ResetPasswordComponent', () => {
  let actRoute: ActivatedRoute;
  let fb: FormBuilder;
  let gigyaService: GigyaService;
  let router: Router;

  beforeEach(() => {
    actRoute = {
      queryParamMap: of({
        get: () => 'pwrt',
      } as Params),
    } as ActivatedRoute;

    fb = {} as FormBuilder;
    fb.group = jest.fn().mockReturnValue({ valid: true, value: {} });

    gigyaService = {} as GigyaService;
    gigyaService.resetPassword = jest.fn().mockImplementation((params: never, callBack: (data: unknown) => void) => {
      callBack({ status: 'OK' })
    });

    router = {} as Router;
    router.navigateByUrl = jest.fn();
  });

  describe('Snapshot', () => {
    beforeEach(async(() => {
      TestBed.configureTestingModule({
        imports: [RouterTestingModule],
        providers: [
          {
            provide: FormBuilder,
            useValue: fb
          },
          {
            provide: GigyaService,
            useValue: gigyaService
          },
          {
            provide: ActivatedRoute,
            useValue: actRoute,
          },
        ],
        declarations: [ResetPasswordComponent],
        schemas: [NO_ERRORS_SCHEMA]
      }).compileComponents();
    }));

    test('check snapshot', () => {
      const fixture = TestBed.createComponent(ResetPasswordComponent);

      expect(fixture).toMatchSnapshot();
    });
  });

  describe('controller', () => {
    it('should set pwrt$', () => {
      return new Promise((resolve) => {
        const sut = new ResetPasswordComponent(
          actRoute,
          fb,
          gigyaService,
          router
        );

        sut.pwrt$.subscribe((val) => {
          expect(val).toEqual('pwrt');
          resolve();
        })

      })
    });
  });

  describe('onSubmit', () => {
    it('should redirect to /login', () => {
      const sut = new ResetPasswordComponent(
        actRoute,
        fb,
        gigyaService,
        router
      );

      sut.onSubmit('token');

      expect(router.navigateByUrl).toBeCalledWith('/login');
    });

    it('should not redirect to /login on gigya error', () => {
      const sut = new ResetPasswordComponent(
        actRoute,
        fb,
        gigyaService,
        router
      );

      gigyaService.resetPassword = jest.fn().mockImplementation((params: never, callBack: (data: unknown) => void) => {
        callBack({ status: 'FAIL', errorDetails: { errors: [true] } })
      });


      sut.onSubmit('token');

      expect(sut.errors).toEqual({ errors: [true] });
      expect(router.navigateByUrl).not.toBeCalledWith('/login');
    });

    it('should not redirect to /login when form is invalid', () => {
      fb.group = jest.fn().mockReturnValue({ valid: false });

      const sut = new ResetPasswordComponent(
        actRoute,
        fb,
        gigyaService,
        router
      );

      sut.onSubmit('token');

      expect(router.navigateByUrl).not.toBeCalledWith('/login');
    });
  });

  describe('onRecover', () => {
    it('should redirect to /login', () => {
      const sut = new ResetPasswordComponent(
        actRoute,
        fb,
        gigyaService,
        router
      );

      sut.onRecover({});

      expect(router.navigateByUrl).toBeCalledWith('/login');
    });
  });
});
