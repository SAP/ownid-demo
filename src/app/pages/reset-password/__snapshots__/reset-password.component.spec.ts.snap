// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordComponent Snapshot check snapshot 1`] = `
<login
  errors="null"
  form={[Function Object]}
  gigyaService={[Function Object]}
  pwrt$={[Function Observable]}
  router={[Function Router]}
>
  <section
    class="sign-body"
  >
    <div
      class="container"
    >
      <div
        class="call-2-action"
      >
        <h1
          class="headline"
        >
           For the best experience use Own Id for login 
        </h1>
        <h2
          class="sub-title"
        >
           Scan QR code with your phone and proceed with out any logins and passwords 
        </h2>
      </div>
      <div>
        <mat-card
          class="sign-section"
        >
          <mat-card-title>
             Set New password or use OwnId to recover your account 
          </mat-card-title>
          
        </mat-card>
      </div>
    </div>
  </section>
</login>
`;
